thirdparty/glib/CMakeLists.txt

```diff
@@ -69,8 +69,8 @@ endif()
 if(NOT DEFINED ENV{DARWIN})
     set(LDFLAGS "${LDFLAGS} -Wl,-rpath-link,${ZLIB_DIR}/lib -Wl,-rpath-link,${LIBFFI_DIR}/.libs -Wl,-rpath-link,${BINARY_DIR}/gmodule/.libs")
 else()
-    set(CFLAGS "-I/usr/local/opt/gettext/include ${CFLAGS}")
-    set(LDFLAGS "${LDFLAGS} -L/usr/local/opt/gettext/lib -L${LIBICONV_DIR}/lib -L${ZLIB_DIR}/lib -L${LIBFFI_DIR}/.libs -L${BINARY_DIR}/gmodule/.libs")
+    set(CFLAGS "-I/usr/local/opt/gettext/include -I/opt/homebrew/include ${CFLAGS}")
+    set(LDFLAGS "${LDFLAGS} -L/usr/local/opt/gettext/lib -L/opt/homebrew/lib -L${LIBICONV_DIR}/lib -L${ZLIB_DIR}/lib -L${LIBFFI_DIR}/.libs -L${BINARY_DIR}/gmodule/.libs")
```

Comment:
I don't like this hardcoding.

Reply:
You mean `/opt/homebrew/lib` should get a variable and become ${HOMEBREW_PREFIX}/lib?

Also is it a good idea to check for brew binary in `Makefile.defs` in case of Darwin host and set the `HOMEBREW_PREFIX` from `$(shell brew --prefix)`? And could this be passed to the lua files somehow? This way custom prefixes set during brew installation could be covered, though probably people who bother to set them know how to fix their builds later on ;)

Reply:
Definitely that, but I'd also prefer to be able to plug in e.g. Nix or gentoo-prefix with little effort if reasonably possible. `HOMEBREW_PREFIX` might allow for that but then the name (at least further down at this point) should really just be something like `MAC_LIBRARY_PREFIX`. But you could always write `MAC_LIBRARY_PREFIX := HOMEBREW_PREFIX` up above in defs or something so it's self-documenting without comments.

Reply:
IIRC, there's a runtime homebrew command that will actually spit out the right prefix for the current system ;).

Reply:
Yup, there's `brew --prefix`. I'll try to umharfcode the paths around evening CET.

What about the paths in lua part, can / should they somehow be based on `brew prefix` as well?

Reply:
Might it be more of a system configuration issue if it doesn't look in `/opt/homebrew/lib` by itself? Put another way, it doesn't seem to be an issue for me on Intel Mac.

Reply:
Basically up above `SDL2` → `libSDL2.dylib` is part of LuaJIT, and then it just checks a bunch of paths for `libSDL2.dylib`. The .so stuff is because `libSDL2-2.0.so` isn't a sane name like `libSDL2.so`. ;-P

Reply:
Yeah, that might be better left to whatever's macOS version of LD_LIBRARY_PATH is (I can't recall ;p).

Reply:
@Frenzie Homebrew keeps linked libs in /usr/local/lib for Intel, which is a common path for tools to search (LD*_PATH is considered insecure for some reason on Darwins). /opt/homebrew was added for apple m1. Here is the descrpition, along with the mention about [ffi ignoring /opt/homebrew/lib until 1.15.2](https://earthly.dev/blog/homebrew-on-m1/#update-third-party-code)? 

>  The problem I encountered was that ffi did not add /opt/homebrew/lib/ to its list of search paths until [version 1.15.2](https://github.com/ffi/ffi/blob/master/CHANGELOG.md#1152--2021-06-16). We were running v1.15.0 ,so every time I tried to run the blog I kept getting no such file or directory errors because ffi wasn’t updated to check /opt/homebrew/lib

I think this might be the reason on the lua side.

Reply:
LuaJIT doesn't use libffi ;).

Reply:
To double check: you have no objections about using `brew --prefix` for MACOS on the Makefile part?

Reply:
`ffi.load` is a thin wrapper around `dlopen`, which will do whatever the system loader is configured (i.e., honor LD_LIBRARY_PATH in the case of GNU ld).

Reply:
> `ffi.load` is a thin wrapper around `dlopen`, which will do whatever the system loader is configured (i.e., honor LD_LIBRARY_PATH in the case of GNU ld).

Agreed for linux, but as [linked before](https://medium.com/macos-is-not-linux-and-other-nix-reflections/d-o-y-ou-ld-library-path-you-6ab0a6135a33), on Darwins you don't rely much on LD_LIBRARY_PATHS

Reply:
I don't care about Apple's inane design/"security" choice. We *heavily* rely on custom search paths (be it via env vars of DT_RPATH entries baked in binaries) ;).

Reply:
Ok, but what of the change in ffi that picks up the `/opt/homebrew/lib` path on its own? I could not find which ffi version is used for koreader, but would it add less complications to the run process to just bump ffi?

Reply:
Aside from the fact that I linked Ruby-FFI issues XD

Reply:
> Homebrew keeps linked libs in /usr/local/lib for Intel, which is a common path for tools to search

Judging by `man dyld` you should try something like `export DYLD_FALLBACK_LIBRARY_PATH="${DYLD_FALLBACK_LIBRARY_PATH}:/opt/homebrew/lib"`.

Reply:
How would you suggest to provide the SDL path to the emulator then? I tried removing the full-path candidates and running  `DYLD_LIBRARY_PATH=/opt/homebrew/lib ./kodev run`, also `export  DYLD_LIBRARY_PATH=/opt/homebrew/lib` and then run, but got 

```
ffi.load: /opt/homebrew/lib/libSDL2.dylib
ffi.load: SDL2
ffi.load (warning): dlopen(libSDL2.dylib, 0x0005): tried: 'libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache), 'libSDL2.dylib' (no such file), '/usr/local/lib/libSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache)
ffi.load: libSDL2-2.0.so
ffi.load (warning): dlopen(libSDL2-2.0.so, 0x0005): tried: 'libSDL2-2.0.so' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibSDL2-2.0.so' (no such file), '/usr/lib/libSDL2-2.0.so' (no such file, not in dyld cache), 'libSDL2-2.0.so' (no such file), '/usr/local/lib/libSDL2-2.0.so' (no such file), '/usr/lib/libSDL2-2.0.so' (no such file, not in dyld cache)
ffi.load: libSDL2-2.0.so.0
ffi.load (warning): dlopen(libSDL2-2.0.so.0, 0x0005): tried: 'libSDL2-2.0.so.0' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibSDL2-2.0.so.0' (no such file), '/usr/lib/libSDL2-2.0.so.0' (no such file, not in dyld cache), 'libSDL2-2.0.so.0' (no such file), '/usr/local/lib/libSDL2-2.0.so.0' (no such file), '/usr/lib/libSDL2-2.0.so.0' (no such file, not in dyld cache)
./luajit: ./ffi/SDL2_0.lua:34: attempt to index local 'SDL' (a boolean value)
```

Reply:
> I could not find which ffi version is used for koreader, but wouldn't it add less complications to the run process to just bump ffi?

None. libffi is like LuaJIT FFI for regular Lua. ;-)

Reply:
the export DYLD_FALLBACK_LIBRARY_PATH also fails to help find libsdl :/

Reply:
> Imposing GNU tools compliance on macs is inane for me, in other words

It's not "GNU tools compliance". It's "hardcoding Homebrew as if Homebrew were the one and only thing of its type that exists is incredibly dumb and I'm nowhere close to accepting such idiocy just yet since at this point I'm not convinced saner avenues have been sufficiently explored."

Reply:
Put another way, the Lua changes are conceptually distinct from the Makefile changes.

Reply:
> > Imposing GNU tools compliance on macs is inane for me, in other words
> 
> It's not "GNU tools compliance". It's "hardcoding Homebrew as if Homebrew were the one and only thing of its type that exists is incredibly dumb and I'm nowhere close to accepting such idiocy just yet since at this point I'm not convinced saner avenues have been sufficiently explored."

man dyld says:
```
 DYLD_FALLBACK_LIBRARY_PATH
              This is a colon separated list of directories that contain
              libraries.  If a dylib is not found at its install  path, dyld
              uses this as a list of directories to search for the dylib.  By
              default, it is set to /usr/local/lib:/usr/lib.
```

On my Ventura all the *_LIBRARY_PATHs I checked are empty. Maybe it's just another time Apple threw in a BSD tool, changed it but didn't bother about documenting in the the manpages, if anywhere?

The only solution I would oppose would be setting global _LIBRARY_ variables in macos, but passing it in a working way (and telling in the build.md to set it to `brew --path` or whatever one uses) looks also better to me than hardcoding the paths. I just have no idea yet how to pass it or make it work ^^

Reply:
So for clarity, after removing all homebrew paths from lua files, the env variables don't seem to do much:

with variables emptied:
```
➜  koreader git:(master) ✗ export DYLD_FALLBACK_LIBRARY_PATH=''        
➜  koreader git:(master) ✗ export DYLD_FALLBACK_LIBRARY_PATH=''
➜  koreader git:(master) ✗ export LD_LIBRARY_PATH=''                  
➜  koreader git:(master) ✗ export DYLD_LIBRARY_PATH=''                
➜  koreader git:(master) ✗ ./kodev run                                    
...
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 01/03/23-22:03:46
 [*] Version: v2022.11-73-gf0b007bb_2023-01-01

ffi.load: libs/libutf8proc.2.dylib
ffi.load: blitbuffer
ffi.load (assisted searchpath): ./libs/libblitbuffer.so
ffi.load: libs/libSDL2.dylib
ffi.load (warning): dlopen(libs/libSDL2.dylib, 0x0005): tried: 'libs/libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibs/libSDL2.dylib' (no such file), '/usr/lib/libs/libSDL2.dylib' (no such file, not in dyld cache), 'libs/libSDL2.dylib' (no such file), '/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file), '/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file)
ffi.load: SDL2
ffi.load (warning): dlopen(libSDL2.dylib, 0x0005): tried: 'libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache), 'libSDL2.dylib' (no such file), '/usr/local/lib/libSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache)
SDL2 not loaded:        ./setupkoenv.lua:27: Not able to load dynamic library: SDL2
./luajit: frontend/device.lua:58: Could not find hardware abstraction for this platform. If you are trying to run the emulator, please ensure SDL is installed.
stack traceback:
        [C]: in function 'error'
        frontend/device.lua:58: in function 'probeDevice'
        frontend/device.lua:61: in main chunk
        [C]: in function 'require'
        ./reader.lua:150: in main chunk
```

with paths provided:
```
➜  koreader git:(master) ✗ export DYLD_FALLBACK_LIBRARY_PATH=/opt/homebrew/lib
➜  koreader git:(master) ✗ export LD_LIBRARY_PATH=/opt/homebrew/lib           
➜  koreader git:(master) ✗ export DYLD_LIBRARY_PATH=/opt/homebrew/lib         
➜  koreader git:(master) ✗ ./kodev run     
...
ffi.load: libs/libSDL2.dylib
ffi.load (warning): dlopen(libs/libSDL2.dylib, 0x0005): tried: 'libs/libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibs/libSDL2.dylib' (no such file), '/usr/lib/libs/libSDL2.dylib' (no such file, not in dyld cache), 'libs/libSDL2.dylib' (no such file), '/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file), '/Users/piotr/Source/koreader/base/build/arm64-apple-darwin22.2.0-debug/libs/libSDL2.dylib' (no such file)
ffi.load: SDL2
ffi.load (warning): dlopen(libSDL2.dylib, 0x0005): tried: 'libSDL2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache), 'libSDL2.dylib' (no such file), '/usr/local/lib/libSDL2.dylib' (no such file), '/usr/lib/libSDL2.dylib' (no such file, not in dyld cache)
SDL2 not loaded:        ./setupkoenv.lua:27: Not able to load dynamic library: SDL2
```

Reply:
otool (or similar ldd-like tools) *probably* have debug knobs to see what the loader is actually doing.

Reply:
At the very least possibly DYLD_PRINT_LIBRARIES=1 (c.f., https://stackoverflow.com/a/51504440)

Reply:
And apparently the hardened runtime disable dyld env vars or... something; but I don't recall how forceful XCode is about that.

Reply:
But SIP probably means if you`exec` from an SIP app, you're essentially fucked? Yay, Apple.

Reply:
So, yeah, seems like a giant "fuck you, FOSS", from Apple. c.f., https://github.com/Homebrew/brew/issues/13481 for more context.

TL;DR: Hardcoded-search paths it is. `>_<".`

Reply:
I think you still meant to base these ones on brew --prefix or some such as well? :-)

Reply:
Yes ;).

Reply:
Ok, just pushed a commit with variables instead of hardcoded brew path ;)